core/trie.go

```diff
@@ -0,0 +1,351 @@
+	if len(pathWords) > 4 {
+		panic("Path too long to fit in Felt")
+	}
+
+	var pathBytes [32]byte
+	for idx, word := range pathWords {
+		startBytes := 24 - (idx * 8)
+		binary.BigEndian.PutUint64(pathBytes[startBytes:startBytes+8], word)
+	}
+
+	pathFelt := felt.NewFelt(0)
+	(&pathFelt).SetBytes(pathBytes[:])
+
+	// https://docs.starknet.io/documentation/develop/State/starknet-state/
+	Pedersen(n.value, &pathFelt, hash)
+	pathFelt.SetUint64(uint64(specPath.Len()))
+	return hash.Add(hash, &pathFelt)
+}
+
+// Equality check of 2 [TrieNode] objects
+func (n *TrieNode) Equal(other *TrieNode) bool {
+	return n.value.Equal(other.value) && n.left.Equal(other.left) && n.right.Equal(n.right)
+}
+
+// Serializes a [TrieNode] in to a byte array
+func (n *TrieNode) MarshalBinary() ([]byte, error) {
+	var ret []byte
+	valueB := n.value.Bytes()
+	ret = append(ret, valueB[:]...)
+
+	if n.left != nil {
+		ret = append(ret, 'l')
+		leftB, err := n.left.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, leftB...)
+	}
+
+	if n.right != nil {
+		ret = append(ret, 'r')
+		rightB, err := n.right.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, rightB...)
+	}
+	return ret, nil
+}
+
+// Deserializes a [TrieNode] from a byte array
+func (n *TrieNode) UnmarshalBinary(data []byte) error {
+	if len(data) < felt.Bytes {
+		return errors.New("Malformed TrieNode bytedata")
+	}
+	n.value = new(TrieValue).SetBytes(data[:felt.Bytes])
+	data = data[felt.Bytes:]
+
+	stream := bytes.NewReader(data)
+	for stream.Len() > 0 {
+		head, err := stream.ReadByte()
+		if err != nil {
+			return err
+		}
+
+		var pathP **StoragePath
+		if head == 'l' {
+			pathP = &(n.left)
+		} else if head == 'r' {
+			pathP = &(n.right)
+		} else {
+			return errors.New("Malformed TrieNode bytedata")
+		}
+
+		*pathP = new(StoragePath)
+		_, err = (*pathP).ReadFrom(stream)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// State [Trie]
+type Trie struct {
+	root    *StoragePath
+	storage TrieStorage
+}
+
+// Converts a [TrieKey] to a [StoragePath] that, when followed on a [Trie] ,leads to the corresponding [TrieNode]
+func PathFromKey(k *TrieKey) *StoragePath {
+	regularK := k.ToRegular()
+	return bitset.FromWithLength(felt.Bits-1, regularK[:])
+}
+
+// Finds the set of common MSB bits in two [StoragePath] objects
+func FindCommonPath(longerPath, shorterPath *StoragePath) (*StoragePath, bool) {
+	divergentBit := uint(0)
+
+	for divergentBit <= shorterPath.Len() &&
+		longerPath.Test(longerPath.Len()-divergentBit) == shorterPath.Test(shorterPath.Len()-divergentBit) {
+		divergentBit++
+	}
+
+	commonPath := shorterPath.Clone()
+	for i := uint(0); i < shorterPath.Len()-divergentBit+1; i++ {
+		commonPath.DeleteAt(0)
+	}
+	return commonPath, divergentBit == shorterPath.Len()+1
+}
+
+// calculates node paths according to the [docs]
+//
+// [docs]: https://docs.starknet.io/documentation/develop/State/starknet-state/
```

Comment:
```suggestion
// GetSpecPath returns the suffix of path that diverges from
// parentPath. For example, for a path 0b1011 and parentPath 0b11,
// this function would return the StoragePath object for 0b10.
//
// This is the canonical representation for paths used in the
// [specification]. Since this trie implementation stores nodes nodes by
// path, we need this function to convert paths to their canonical
// representation.
//
// [specification]: https://docs.starknet.io/documentation/develop/State/starknet-state/
```